### PR TITLE
Issue25483 pipeline options refactoring

### PR DESCRIPTION
**Pull request by Malarg**
resolves: #25483
pipeline options extracted to playground_components and added to tour of beam

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_dropdown_separator.dart [hunk 1]
```diff
@@ -17,16 +17,17 @@
  */
 
 import 'package:flutter/material.dart';
-import 'package:playground/constants/sizes.dart';
-import 'package:playground_components/playground_components.dart';
+
+import '../../constants/sizes.dart';
+import '../../theme/theme.dart';
 
 class PipelineOptionsDropdownSeparator extends StatelessWidget {
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_dropdown_separator.dart:24]: Can we reuse `BeamDivider` for this?

File: playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_row.dart [hunk 2]
```diff
@@ -57,13 +57,13 @@ class PipelineOptionsRow extends StatelessWidget {
...
+          width: BeamIconSizes.large,
           child: IconButton(
-            iconSize: kIconSizeMd,
-            splashRadius: kIconButtonSplashRadius,
+            iconSize: BeamIconSizes.medium,
+            splashRadius: BeamIconSizes.largeSplashRadius,
             icon: const Icon(
               Icons.delete_outlined,
-              color: kLightPrimary,
+              color: BeamLightThemeColors.primary, //Color(0xFFE74D1A)
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_row.dart:66]: How about the dark theme?

File: learning/tour-of-beam/frontend/lib/components/scaffold.dart [hunk 3]
```diff
@@ -21,21 +21,21 @@ import 'package:flutter/material.dart';
...
 import 'login/button.dart';
 import 'logo.dart';
-import 'pipeline_options/button.dart';
 import 'profile/avatar.dart';
 import 'sdk_dropdown.dart';
 
 class TobScaffold extends StatelessWidget {
   final Widget child;
-  final PlaygroundController? playgroundController;
+  final TourNotifier? tourNotifier;
```
[the remark this answers is not in the document]
Comment (Malarg) [reply learning/tour-of-beam/frontend/lib/components/scaffold.dart]: fixed

File: playground/frontend/playground_components/lib/src/constants/constants.dart [hunk 4]
```diff
@@ -17,3 +17,5 @@
  */
 
 const spaceCount = 2;
+
+const double kLabelFontSize = 16.0;
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/constants/constants.dart]: fixed

File: playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_dropdown.dart [hunk 5]
```diff
@@ -16,10 +16,10 @@
  * limitations under the License.
  */
 
+import 'package:easy_localization/easy_localization.dart';
 import 'package:flutter/material.dart';
-import 'package:flutter_gen/gen_l10n/app_localizations.dart';
-import 'package:playground/modules/editor/components/pipeline_options_dropdown/pipeline_options_dropdown_body.dart';
-import 'package:playground_components/playground_components.dart';
+import '../dropdown_button/dropdown_button.dart';
```
[the remark this answers is not in the document]
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_dropdown.dart:22]: done

File: playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_dropdown_separator.dart (hunk 1 above)
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_dropdown_separator.dart]: Yes, done

File: playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_row.dart (hunk 2 above)
Comment (Malarg) [reply playground/frontend/playground_components/lib/src/widgets/pipeline_options_dropdown/pipeline_options_row.dart]: Correct, done

File: learning/tour-of-beam/frontend/lib/pages/tour/widgets/pipeline_options.dart [hunk 6]
```diff
@@ -19,26 +19,35 @@
...
-  final VoidCallback close;
-  final String pipelineOptions;
+import '../state.dart';
 
-  const PipelineOptionsDropdownBody({
-    required this.close,
-    required this.pipelineOptions,
-  });
+class TobPipelineOptionsDropdown extends StatelessWidget {
+  final TourNotifier tourNotifier;
```
Comment (alexeyinkin) [inline learning/tour-of-beam/frontend/lib/pages/tour/widgets/pipeline_options.dart:25]: PlaygroundController is sufficient for this.

File: learning/tour-of-beam/frontend/lib/pages/tour/widgets/playground_demo.dart [hunk 7]
```diff
@@ -66,4 +74,15 @@ class PlaygroundDemoWidget extends StatelessWidget {
...
+
+  void _handleError(BuildContext context, PlaygroundController controller) {
+    PlaygroundComponents.toastNotifier.add(
+      Toast(
+        description: controller.codeRunner.result?.errorMessage ?? '',
+        title: 'intents.playground.runCode'.tr(),
+        type: ToastType.error,
+      ),
+    );
+    controller.resetErrorMessageText();
```
Comment (alexeyinkin) [inline learning/tour-of-beam/frontend/lib/pages/tour/widgets/playground_demo.dart:86]: Check if there is an issue to refactor this. Link in a TODO.

File: playground/frontend/playground_components/lib/src/constants/sizes.dart [hunk 8]
```diff
@@ -39,11 +39,14 @@ class BeamSizes {
 
   static const double appBarHeight = 55;
   static const double buttonHeight = 40;
+  static const double dividerHeight = 1;
   static const double elevation = 2;
   static const double headerButtonHeight = 46;
+  static const double labelFontSize = 16;
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/constants/sizes.dart:45]: Delete.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: learning/tour-of-beam/frontend/lib/pages/tour/screen.dart [hunk 9]
```diff
@@ -35,7 +36,11 @@ class TourScreen extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     return TobScaffold(
-      playgroundController: tourNotifier.playgroundController,
+      pageActions: [
+        TobPipelineOptionsDropdown(
+          playgroundController: tourNotifier.playgroundController,
```
Comment (alexeyinkin) [inline learning/tour-of-beam/frontend/lib/pages/tour/screen.dart:41]: Put `TobScaffold` into AnimatedBuilder. Pass `null` for `playgroundController` if there is no snippet for the unit.

File: playground/frontend/playground_components/lib/src/controllers/playground_controller.dart [hunk 10]
```diff
@@ -141,6 +147,15 @@ class PlaygroundController with ChangeNotifier {
     );
   }
 
+  void setEmpty() {
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/controllers/playground_controller.dart:150]: Delete.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: learning/tour-of-beam/frontend/lib/pages/tour/state.dart [hunk 11]
```diff
@@ -37,7 +37,7 @@ import 'path.dart';
 
 class TourNotifier extends ChangeNotifier with PageStateMixin<void> {
   final ContentTreeController contentTreeController;
-  final PlaygroundController playgroundController;
+  PlaygroundController? playgroundController;
```
Comment (alexeyinkin) [inline learning/tour-of-beam/frontend/lib/pages/tour/state.dart:40]: TourNotifier must always have a `final PlaygroundController`.
When I talked nullable, I meant that `TobScaffold.playgroundController` should be `null`.

File: learning/tour-of-beam/frontend/lib/components/scaffold.dart
Comment (alexeyinkin) [inline learning/tour-of-beam/frontend/lib/components/scaffold.dart:1]: Revert please.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: playground/frontend/playground_components/lib/src/controllers/playground_controller.dart [hunk 12]
```diff
@@ -35,6 +35,7 @@ import '../models/example_loading_descriptors/user_shared_example_loading_descri
 import '../models/intents.dart';
 import '../models/sdk.dart';
 import '../models/shortcut.dart';
+import '../models/snippet_file.dart';
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/controllers/playground_controller.dart:38]: Unused.

File: playground/frontend/playground_components/lib/src/controllers/example_loaders/examples_loader.dart [hunk 13]
```diff
@@ -53,6 +53,10 @@ class ExamplesLoader {
     _playgroundController = value;
   }
 
+  void clearDescriptor() {
+    _descriptor = null;
+  }
+
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/controllers/example_loaders/examples_loader.dart:59]: Delete.

File: learning/tour-of-beam/frontend/lib/components/scaffold.dart [hunk 14]
```diff
@@ -115,3 +119,38 @@ class _SdkSelector extends StatelessWidget {
...
+        );
+        return Row(
+          children: widgets
+              .map(
+                (e) => Padding(
+                  padding: const EdgeInsets.only(left: BeamSizes.size12),
+                  child: e,
+                ),
+              )
+              .toList(),
```
Comment (alexeyinkin) [inline learning/tour-of-beam/frontend/lib/components/scaffold.dart:151]:
```suggestion
              .toList(growable: false),
```

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

**Review by alexeyinkin — APPROVED** [review]
LGTM (internal review).

**Pull request merged**